pennylane/gradients/parameter_shift.py

```diff
@@ -592,7 +629,7 @@ def param_shift(
             "If this is unintended, please mark trainable parameters in accordance with the "
             "chosen auto differentiation framework, or via the 'tape.trainable_params' property."
         )
-        return [], lambda _: ()
+        return [], lambda _: np.zeros((tape.output_dim, 0))

```

Comment:
If I'm not mistaken, `param_shift` is a somewhat more low-level function than `qml.grad` and `qml.jacobian`. For example, the derivatives for parameters that are not included in `argnum`, if given, are set to `0` instead of being skipped by `param_shift`, whereas `qml.grad` would not return any object/entry for such parameters...

Reply:
Ah I see now. Well `argnum` behaves a bit strangely for the gradient transforms, as it only allows to select/deselect arguments that are already in the `tape.trainable_params`. But the "true" number of trainable tape parameters does indeed directly determine the last dimension of the output gradient array, so returning an empty array when there are no trainable parameters is the natural extension here 👍 
In this case, would mind changing this line for the other gradient transforms as well (finite-diff, cv)? 

Although practically, I'm not sure the dimension information in empty return array with some is of any use. Do you know of some? :) There is also the problem that `tape.output_dim` only contains the "flattened" tape output dimensions (e.g. it will not give the correct dimensions if the tape contains `qml.probs(0)` and `qml.probs(1)`, a `(2, 2)` output shape).

Reply:
>Ah I see now. Well argnum behaves a bit strangely for the gradient transforms, as it only allows to select/deselect arguments that are already in the tape.trainable_params.

Yes, I feel like we really should work a bit more towards a consistent system, also regarding wording, across the codebase (also referring to our chat about `argnum`). 

> In this case, would you mind changing this line for the other gradient transforms as well (finite-diff, cv)?

Done :)

>Although practically, I'm not sure the dimension information in the empty return array is of any use. Do you know of some? 

That is a fair point. I am not sure, but I personally thought of homogeneous return types being more convenient in general. Something like `hessian.shape` would work even if there are no trainable parameters, for example :)

>There is also the problem that tape.output_dim only contains the "flattened" tape output dimensions

That's a good point! :thinking: I suppose for now we may stick with this though? Is there a plan to adapt `tape.output_dim` to such non-flat cases?

Reply:
>That's a good point! 🤔 I suppose for now we may stick with this though? Is there a plan to adapt tape.output_dim to such non-flat cases?

Yes I believe this PR is working on it actually https://github.com/PennyLaneAI/pennylane/pull/2044 :)